src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1417 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.Json;
+using Microsoft.IdentityModel.Json.Linq;
+using Microsoft.IdentityModel.JsonWebTokens;
+using Microsoft.IdentityModel.Logging;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.IdentityModel.Protocols.SignedHttpRequest
+{
+    /// <summary>
+    /// A handler designed for creating and validating signed http requests. 
+    /// </summary>
+    /// <remarks>The handler implementation is based on 'A Method for Signing HTTP Requests for OAuth' specification.</remarks>
+    public class SignedHttpRequestHandler
+    {
+        // (https://tools.ietf.org/html/draft-ietf-oauth-signed-http-request-03#section-3.2)
+        // "Encodes the name and value of the header as "name: value" and appends it to the string buffer separated by a newline "\n" character."
+        private readonly string _newlineSeparator = "\n";
+
+        private readonly JsonWebTokenHandler _jwtTokenHandler = new JsonWebTokenHandler();
+        private readonly Uri _baseUriHelper = new Uri("http://localhost", UriKind.Absolute);
+        private readonly HttpClient _defaultHttpClient = new HttpClient();
+
+        #region SignedHttpRequest creation
+        /// <summary>
+        /// Creates a signed http request using the <paramref name="signedHttpRequestDescriptor"/>.
+        /// /// </summary>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <param name="cancellationToken">Propagates notification that operations should be canceled.</param>
+        /// <returns>A signed http request as a JWS in Compact Serialization Format.</returns>
+        public async Task<string> CreateSignedHttpRequestAsync(SignedHttpRequestDescriptor signedHttpRequestDescriptor, CancellationToken cancellationToken)
+        {
+            if (signedHttpRequestDescriptor == null)
+                throw LogHelper.LogArgumentNullException(nameof(signedHttpRequestDescriptor));
+
+            var header = CreateHttpRequestHeader(signedHttpRequestDescriptor);
+            var payload = CreateHttpRequestPayload(signedHttpRequestDescriptor);
+            return await SignHttpRequestAsync(header, payload, signedHttpRequestDescriptor, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Creates a JSON representation of a HttpRequest header.
+        /// </summary>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <returns>A JSON representation of an HttpRequest header.</returns>
+        protected virtual string CreateHttpRequestHeader(SignedHttpRequestDescriptor signedHttpRequestDescriptor)
+        {
+            var header = new JObject
```

Comment:
Let JsonWebTokenHandler create the header

Reply:
I'll refactor the code, it makes sense to utilize JsonWebTokenHandler. The method will be synchronous for now, but we can add another async method when we support async in JsonWebTokenHandler.

Reply:
Addressed with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1260/commits/a1656abf4f653567a5d6e97a49aeed2559d331a5